README

DEBIAN PACKAGE POPULARITY CONTEST - Avery Pennarun <apenwarr@debian.org>
=================================

This package contains a script, /usr/sbin/popularity-contest, which
generates a list of the packages installed on your system, in order of
most-recently-used to least-recently-used.  The simplest way to use this
information is to help clean up your hard drive by removing unused packages.

For example,
	popularity-contest | grep '<OLD>'
will show you a list of packages you haven't used in a while.  Note that
this output isn't totally accurate: some packages appear "old" but you can't
remove them because other (non-old) packages depend on them.  Shared library
packages are particularly bad this way because it's impossible to tell when
a library was last used.

The popularity-contest output looks like this:

  POPULARITY-CONTEST-0 TIME:914183330 ID:b92a5fc1809d8a95a12eb3a3c84166dd
  914183333 909868335 grep /bin/fgrep
  914183333 909868280 findutils /usr/bin/find
  914183330 909885698 dpkg-awk /usr/bin/dpkg-awk
  914183330 909868577 gawk /usr/bin/gawk
  [...more lines...]
  END-POPULARITY-CONTEST-0 TIME:914183335

The first and last lines allow you to put more than one set of
popularity-contest results into a single file and then split them up
easily later.

The rest of the lines are package entries, one line for each package
installed on your system.  They have the format:

	<atime> <ctime> <package-name> <mru-program> <tag>

<package-name> is the name of the Debian package that contains <mru-program>.
<mru-program> is the most recently used program, static library, or header
(.h) file in the package.

<atime> and <ctime> are the access time and creation time of the
<mru-program> on your disk, respectively, represented as the number of
seconds since midnight GMT on January 1, 1970 (i.e. in Unix time_t format).
Linux updates <atime> whenever you open the file; <ctime> was set when you
first installed the package.

<tag> is determined by popularity-contest depending on <atime>, <ctime>, and
the current date.  <tag> can be RECENT-CTIME, OLD, or NOFILES.

RECENT-CTIME means that atime is very close to ctime; it's impossible to
tell whether the package was used recently or not, since <atime> is also
updated when <ctime> is set.  Normally, this happens because you have
recently upgraded the package to a new version, resetting the <ctime>.

OLD means that the <atime> is more than a month ago; you haven't used the
package for more than a month.

NOFILES means that no files in the package seemed to be programs, so
<atime>, <ctime>, and <mru-program> are invalid.


PARTICIPATING IN THE PUBLIC SURVEY
==================================

You can help out the Debian Project by sending your popularity-contest
results to us anonymously.  With information from enough people, we can
determine which packages are most often used, which are often installed but
not used, and which are nearly never used.  That way, we can help people
make decisions about which packages should fit on the first Debian install
CD, for example, or which packages are most important for an FTP mirror site
to keep.

  !!!!

SECURITY NOTE: it's impossible to make a submission completely anonymous,
since Internet servers tend to add headers and log messages along the way.
Our receiver program at debian throws away this information as soon as
possible so no one will see it, but if you're really paranoid you might not
want to participate.

  !!!!

The popularity-contest package comes with a cron task to send the
information out to us each week.  You can change where the submissions go
by editing /etc/popularity-contest.conf.  Normally, you will send the
results to:
		survey@popcon.debian.org
	
The result of the survey is available at

		http://popcon.debian.org/

You can find some sample analysis scripts in
/usr/share/doc/popularity-contest as part of this package.

RELATED PROJECTS
================

There are several projects trying to count users and machines.  First and
foremost is the Linux Counter at http://counter.li.org/.  It tries to
estimate the amount of Linux users worldwide.

There are some popcon installations for other distributions:

Ubuntu: http://popcon.ubuntu.com/
DebianEdu: http://popcon.skolelinux.org/
NeuroDebian: http://neuro.debian.net/popcon/

INSTALLATION PRESEEDING
=======================

The package supports preseeding during installation to change the URL
used to submit reports.  The template name is
'popularity-contest/submiturls', and it expects one or more space
separated URLs to use when submitting.

QUESTIONS ?
===========

Please look at the FAQ in /usr/share/doc/popularity-contest/FAQ
or for an up-to-date version, online at http://popcon.debian.org/FAQ


FINDING THE SOURCE
==================

This package is being maintained in Subversion using Alioth.  The project summary
page is available from <URL:https://alioth.debian.org/projects/popcon/>.
The project home page is at <URL:http://popcon.alioth.debian.org/>.

REPORTING BUGS
==============

Report bugs into the Debian Bug Tracking System.  See
<URL:http://bugs.debian.org/> for instructions.  If you want to get in
touch with the popularity contest maintainers, send an email to
popcon-developers@lists.alioth.debian.org.


debian/changelog
popularity-contest (1.65) UNRELEASED; urgency=low

  * popanal.py, gensections.pl, popcon.pl:
    - Read Packages.xz instead of Packages.gz.

 -- Bill Allombert <ballombe@debian.org>  Fri, 24 Jun 2016 15:43:16 +0200

popularity-contest (1.64) unstable; urgency=low

  * debian/dirs:
    - ship /etc/cron.d. Closes: #798941
  * debian/postrm:
    - remove /etc/cron.d/popularity-contest on purge

 -- Bill Allombert <ballombe@debian.org>  Tue, 15 Sep 2015 16:30:57 +0200

popularity-contest (1.63) unstable; urgency=low

  * Thanks goes to Peter Palfrader for fixing popcon CGI performance
    issues.
  * debian/cron.daily:
   - Only update /var/log/popcon if submission was successful
   - Add support for --crond option
  * debian/rules:
    - use dh_prep instead of dh_clean -k
  * debian/postinst:
   - Generate a file /etc/cron.d/popularity-contest to randomize
     the submission time to avoid overloading the server. Closes: #709892
   - Do not hard-code path to uuidgen.
  * examples/cgi-bin/popcon.cgi:
    - document that it requires libcgi-pm-perl due to change in perl.
  * Patches from Paul Wise:
   - debian/changelog: fix some typos
   - examples/cgi-bin/popcon.cgi: fix typo in comments
   - README, debian/control, debian/po/*.po: remove trailing whitespaces
   - debian/control: Vcs fields: use https URL.

 -- Bill Allombert <ballombe@debian.org>  Sun, 13 Sep 2015 21:08:47 +0200

popularity-contest (1.62) unstable; urgency=low

  * Please note when reading stats that Jessie was released with 1.61.
    Closes: #783502
  * debian/control:
   + Updated Standards-Version from 3.9.5 to 3.9.6.  No change needed.
   + Remove dependency on libio-socket-ip-perl (>= 0.25-3) included
     in perl-base.
   + Vcs-Git, Vcs-Browser: migrate to anonscm.debian.org URL
   + Recommends: cron-daemon instead of fcron. Closes: #752504
     Thanks Alexandre Detiste.
  * debian/rules:
   + Remove obsolete 'source' and 'diff' targets.
  * prepop.pl: reject reports with invalid ARCH or POPCONVER field.
    Such fields are included in the HTML report without quoting,
    so a malicious submission could inject content in the HTML report
    Thanks Paul Wise.
  * Update example server-side scripts to popcon.d.o version.
  * debian/cron.daily:
    - use runuser instead of su. Closes: #764369. Thanks Laurent Bigonville
  * debian-popcon.gpg: use new submission key

 -- Bill Allombert <ballombe@debian.org>  Mon, 27 Apr 2015 17:10:46 +0200

popularity-contest (1.61) unstable; urgency=low

  * Add support for IPv6. Closes: #730760. Thanks Bob Ham.
   + popcon-upload:
     - Use IO::Socket::IP instead of IO::Socket::INET in popcon-upload.
   + debian/control:
     - Add a dependency on libio-socket-ip-perl (>= 0.25-3).
       This module is targeted for inclusion in perl-base.
  * examples/bin/popcon.pl:
   - New popcon.debian.org layout by Stéphane Blondon
  * debian/control:
   - Updated Standards-Version from 3.9.4 to 3.9.5.  No change needed.
   - Add Vcs-Browser and Vcs-Git fields. Closes: #729024.
     Thanks Yaroslav Halchenko.
   - Petter has retired as popcon maintainer. Adjust Uploaders field.
  * Allow derivatives to set up their own public keys in addition to Debian.
    Closes: #729097. Thanks Yaroslav Halchenko.
   + debian/cron.daily:
     - Add support for /etc/popularity-contest.d/
     - Allow multiple keyrings and keys to be set.
   + FAQ: Add section for derivative distributions
     - Document /etc/popularity-contest.d/

 -- Bill Allombert <ballombe@debian.org>  Mon, 17 Mar 2014 20:37:27 +0100

popularity-contest (1.60) unstable; urgency=low

  * Encrypt reports by default:
    + default.conf, FAQ:
      - Set ENCRYPT=maybe by default and document it.
    As a consequence, reports are now compressed by default. Closes: #149425

 -- Bill Allombert <ballombe@debian.org>  Thu, 05 Sep 2013 09:29:57 +0200

popularity-contest (1.59) unstable; urgency=low

  * debian/cron.daily:
    - Provide a temporary homedir to gpg so it does not create /root/.gnupg.
      Also use 'set -e'. Closes: #714917 Thanks Ansgar Burchardt.
    - Add setting ENCRYPT=maybe, which encrypt if gpg is available.
      Make ENCRYPT=yes to fail if gpg is not available. Closes: #714918

 -- Bill Allombert <ballombe@debian.org>  Thu, 01 Aug 2013 16:26:31 +0200

popularity-contest (1.58) unstable; urgency=low

  * Move examples script to examples subdirectory.
    - Add gensections.pl
    - Do not install clean-filter/clean-genpkglist
    - Rename popcon-submit.cgi to popcon.cgi
  * popcon now report the dpkg Vendor field. Suggested by Paul Wise.
  * popanal.py:

prepop.pl
#!/usr/bin/perl -wT
# Accept popularity-contest entries on stdin and drop them into a
# subdirectory with a name based on their MD5 ID.
#
# Only the most recent entry with a given MD5 ID is kept.
#

$dirname = 'popcon-entries';
$now = time;
$state='initial'; # one of ('initial','accept','reject')

my($file,$mtime);
while(<>)
{
    $state eq 'initial' and do
    {
       /^POPULARITY-CONTEST-0/ or next;
       my @line=split(/ +/);
       my %field;
       for (@line)
       {
	    my ($key, $value) = split(':', $_, 2);
            $field{$key}=$value;
       };
       $id=$field{'ID'};
       if (!defined($id) || $id !~ /^([a-f0-9]{32})$/) 
       {
         print STDERR "Bad hostid: $id\n";
         $state='reject'; next;
       }
       $id=$1; #untaint $id
       $mtime=$field{'TIME'};
       if (!defined($mtime) || $mtime!~/^([0-9]+)$/)
       {
         print STDERR "Bad mtime $mtime\n";
         $state='reject'; next;
       }
       $mtime=int $1; #untaint $mtime;
       $mtime=$now if ($mtime > $now);
       my $dir=substr($id,0,2);
       unless (-d "$dirname/$dir") {
         mkdir("$dirname/$dir",0755) or do {$state='reject';next;};
       };
       $file="$dirname/$dir/$id"; 
       open REPORT, ">",$file or do {$state='reject';next;};
       print REPORT $_;
       $state='accept'; next;
    };
    $state eq 'reject' and do
    {
      /^From/ or next;
      $state='initial';next;
    };
    $state eq 'accept' and do
    {
      /^From/ and do 
      {
        close REPORT; 
        unlink $file; 
        print STDERR "Bad report $file";
        $state='initial';
        next;
      };
      print REPORT $_; #accept line.
      /^END-POPULARITY-CONTEST-0/ and do 
      {
        close REPORT; 
        utime $mtime, $mtime, $file;
        $state='initial';
        next;
      };
    };
}
if ($state eq 'accept')
{
        close REPORT;
        unlink $file; #Reject
        print STDERR "Bad report $file";
}


popcon-process.sh
#!/bin/sh

BASEDIR=/org/popcon.debian.org/popcon-mail
MAILDIR=../Mail
WEBDIR=../www
LOGDIR=$BASEDIR/../logs
BINDIR=$BASEDIR/../bin
DATADIR=$BASEDIR/popcon-entries
SUMMARYDIR=$BASEDIR/all-popcon-results
SUMMARYDIRSTABLE=$BASEDIR/all-popcon-results.stable

# set to 'true' if email submissions should be processed
READMAIL=true

# Remove entries older than # number of days
DAYLIMIT=20

set -e
cd $BASEDIR
umask 0002

# rotate incoming mail spool files
if [ true = "$READMAIL" ] ; then
    mv $MAILDIR/survey new-popcon-entries
    touch $MAILDIR/survey
    chmod go-rwx $MAILDIR/survey

    # process entries, splitting them into individual reports
    $BINDIR/prepop.pl <new-popcon-entries >$LOGDIR/prepop.out 2>&1
fi

# delete outdated entries
rm -f results results.stable
find $DATADIR -type f -mtime +$DAYLIMIT -print0 | xargs -0 rm -f --

# Generate statistics
find $DATADIR -type f | xargs cat \
        | nice -15 $BINDIR/popanal.py >$LOGDIR/popanal.out 2>&1
cp results $WEBDIR/all-popcon-results
cp results.stable $WEBDIR/stable/stable-popcon-results
gzip -f $WEBDIR/all-popcon-results
gzip -f $WEBDIR/stable/stable-popcon-results
cp $WEBDIR/all-popcon-results.gz $SUMMARYDIR/popcon-`date +"%Y-%m-%d"`.gz
cp $WEBDIR/stable/stable-popcon-results.gz $SUMMARYDIRSTABLE/popcon-`date +"%Y-%m-%d"`.stable.gz

cd ../popcon-stat
find $SUMMARYDIR -type f -print | sort | $BINDIR/popcon-stat.pl ../www/stat>$LOGDIR/popstat.log 2>&1 
find $SUMMARYDIRSTABLE -type f -print | sort | $BINDIR/popcon-stat.pl ../www/stable/stat >> $LOGDIR/popstat.log 2>&1 

cd ../popcon-web
$BINDIR/popcon.pl >$LOGDIR/popcon.log 2>$LOGDIR/popcon.errors


popcon.pl
#! /usr/bin/perl -wT

use strict;

$ENV{PATH}="/bin:/usr/bin";

my %results=('all' => "../popcon-mail/results", 'stable' => "../popcon-mail/results.stable");
my $popbase = "../www";
my %popcon= ('all' => "", 'stable' => "/stable");
my %popfile=('all' => "all-popcon-results.gz", 'stable' => "stable-popcon-results.gz");
my %poptext=('all' => "All reports", 'stable' => "Stable reports");
my $mirrorbase = "/srv/mirrors/debian";
my $docurlbase = "/";
my %popconver=("1.28" => "sarge", "1.41" => "etch", "1.46" => "lenny", "1.49" => "squeeze");
my %popver=();
my @dists=("main","contrib","non-free","non-US");
my @fields=("inst","vote","old","recent","no-files");
my %maint=();
my %list_header=(
"maint" => <<"EOF",
#<name> is the developer name;
#
#The fields below are the sum for all the packages maintained by that
#developer:
EOF
"source" => <<"EOF",
#<name> is the source package name;
#
#The fields below are the sum for all the binary packages generated by
#that source package:
EOF
"sourcemax" => <<"EOF");
#<name> is the source package name;
#
#The fields below are the maximum for all the binary packages generated by
#that source package:
EOF

# Progress indicator

sub mark
{
  print join(" ",$_[0],times),"\n";
}

# HTML templates

sub htmlheader
{
  print HTML <<"EOH";
  <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
  <html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
      <title> Debian Popularity Contest </title>
        <link rev="made" href="mailto:ballombe\@debian.org">
        <link rel="shortcut icon" href="/favicon.ico">
        </head>
        <body text="#000000" bgcolor="#FFFFFF" link="#0000FF" vlink="#800080" alink="#FF0000">
        <div align="center">
        <a href="http://www.debian.org/">
        <img src="http://www.debian.org/logos/openlogo-nd-50.png" border="0" hspace="0" vspace="0" alt="" width="50" height="61">
        </a>
        <a href="http://www.debian.org/">
        <img src="http://www.debian.org/Pics/debian.jpg" border="0" hspace="0" vspace="0" alt="Debian Project" width="179" height="61">
        </a>
        </div>
        <br>
        <table bgcolor="#DF0451" border="0" width="100%" cellpadding="0" cellspacing="0" summary="">
        <tr>
        <td valign="top">
        <img src="http://www.debian.org/Pics/red-upperleft.png" align="left" border="0" hspace="0" vspace="0" alt="" width="15" height="16">
        </td>
        <td rowspan="2" align="center">
        <font color="#FFFF00"><big><big>Debian Popularity Contest</big></big></font>
        </td>
        <td valign="top">
        <img src="http://www.debian.org/Pics/red-upperright.png" align="right" border="0" hspace="0" vspace="0" alt="" width="16" height="16">
        </td>
        </tr>
        <tr>
        <td valign="bottom">
        <img src="http://www.debian.org/Pics/red-lowerleft.png" align="left" border="0" hspace="0" vspace="0" alt="" width="16" height="16">
        </td>
        <td valign="bottom">
        <img src="http://www.debian.org/Pics/red-lowerright.png" align="right" border="0" hspace="0" vspace="0" alt="" width="15" height="16">
        </td>
        </tr>
        </table>
EOH
}

sub popconintro
{
  my ($name,$page) = @_;
  &htmlheader;
  print HTML <<"EOF";
  <p> <em> The popularity contest project is an attempt to map the usage of
  Debian packages.  This site publishes the statistics gathered from report
  sent by users of the <a
  href="http://packages.debian.org/popularity-contest">popularity-contest</a>
  package. This package sends every week the list of packages installed and the
  access time of relevant files to the server via email. Every day the server
  anonymizes the result and publishes this survey.
  For more information, read the <a href="${docurlbase}README">README</a> and the 
  <a href="${docurlbase}FAQ">FAQ</a>.
  </em> <p>
<form method="GET" action="http://qa.debian.org/popcon.php">Popcon statistics
for source package <input type="text" size="30" maxlength="80" name="package">
<input type="submit" value="Go">
</form> <p>

<style type="text/css">
  #tabs ul { padding: 0; margin: 0; background #DF0451; }
  #tabs li {
    display: inline;
    border: 2px #DF0451 solid;
    border-bottom-width: 0;
    margin: 0 2px 0 0;
    font-size: 140%;
    padding: 0 2px;
    -moz-border-radius: 15px 15px 0 0; border-radius: 15px 15px 0 0;
  }
  #tabs #current { background: #DF0451; color: #FFFF00; }
  #main { border: 2px #DF0451 solid; 
  -moz-border-radius: 0 15px 15px 15px; border-radius: 0 15px 15px 15px; }
</style>
<div id="tabs">
  <ul>
EOF
  for (keys %poptext)
  {
    if ($_ eq $name) {
      print HTML "<li id=\"current\">$poptext{$_}</li>\n";
    } else {
      print HTML "<li><a href=\"$popcon{$_}/$page\">$poptext{$_}</a></li>\n";
    }
  }
  print HTML <<"EOF";
  </ul>
</div>
<div id="main">
EOF
}

sub htmlfooter
{
  my ($numsub) = @_;
  my $date=gmtime();
  print HTML <<EOF;
<pre>
inst     : number of people who installed this package;
vote     : number of people who use this package regularly;
old      : number of people who installed, but don't use this package regularly;
recent   : number of people who upgraded this package recently;
no-files : number of people whose entry didn't contain enough information (atime
and ctime were 0).
</pre>
<p>
Number of submissions considered: $numsub
</p><p>
To participate in this survey, install the <a href="http://packages.debian.org/popularity-contest">popularity-contest</a> package.
</p>
EOF
  print HTML <<EOH
<p>
</div>
<small>
Made by <a href="mailto:ballombe\@debian.org"> Bill Allombert </a>. Last generated on $date UTC. <br>
<a href="http://popcon.alioth.debian.org" > Popularity-contest project </a> by Avery Pennarun, Bill Allombert and Petter Reinholdtsen.
<BR>
Copyright (C) 2004-2005 <A HREF="http://www.spi-inc.org/">SPI</A>;
See <A HREF="http://www.debian.org/license">license terms</A>.
</small>
</body>
</html>
EOH
}

# Report generators

sub make_sec
{
  my $sec="$_[0]/$_[1]";
  -d $sec || system("mkdir","-p","$sec");
}

sub print_by
{
   my ($dir,$f)=@_;
   print HTML ("<a href=\"$dir/by_$f\">$f</a> [<a href=\"$dir/by_$f.gz\">gz</a>] ");
}

sub make_by
{
  my ($popcon,$sec,$order,$pkg,$winner,$listp) = @_;
  my (%sum, $me);
  my @list = sort {$pkg->{$b}->{$order}<=> $pkg->{$a}->{$order} || $a cmp $b } @{$listp};
  $winner->{"$sec/$order"}=$list[0];
  open DAT , "|-:utf8", "tee $popcon/$sec/by_$order | gzip -c > $popcon/$sec/by_$order.gz";
  if (defined($list_header{$sec}))
  {
    print DAT $list_header{$sec};
    $me="";
  }
  else 
  {
    print DAT <<"EOF";
#Format
#   
#<name> is the package name;
EOF
    $me="(maintainer)";
  }
  print DAT << "EOF";
#<inst> is the number of people who installed this package;
#<vote> is the number of people who use this package regularly;
#<old> is the number of people who installed, but don't use this package
#      regularly;
#<recent> is the number of people who upgraded this package recently;
#<no-files> is the number of people whose entry didn't contain enough
#           information (atime and ctime were 0).
#rank name                            inst  vote   old recent no-files $me
EOF
  my $format="%-5d %-30s".(" %5d"x($#fields+1))." %-32s\n";
  my $rank=0;
  my $p;
  for $p (@list)
  {
    $rank++;
    my $m=(defined($list_header{$sec})?"":"($maint{$p})");
    printf  DAT $format, $rank, $p, (map {$pkg->{$p}->{$_}} @fields), $m;
    $sum{$_}+=$pkg->{$p}->{$_} for (@fields);
  }
  print  DAT '-'x66,"\n";
  printf DAT $format, $rank, "Total", map {defined($sum{$_})?$sum{$_}:0} @fields, "";
  close DAT;
}

sub make
{
  my ($popcon, $sec,$pkg,$winner,$list)=@_;
  make_sec ($popcon,$sec);
  make_by ($popcon, $sec, $_, $pkg, $winner, $list) for (@fields);
}
sub print_pkg
{
  my ($pkg)=@_;
  return unless (defined($pkg));
  my $size=length $pkg;
  my $pkgt=substr($pkg,0,20);
  print HTML "<a href=\"http://packages.debian.org/$pkg\">$pkgt</a> ",
  ' 'x(20-$size);
}

my %section=();
my %source=();

#Format
#<name> <vote> <old> <recent> <no-files>
#   
#<name> is the package name;
#<vote> is the number of people who use this package regularly;
#<old> is the number of people who installed, but don't use this package
#        regularly;
#<recent> is the number of people who upgraded this package recently;
#<no-files> is the number of people whose entry didn't contain enough
#        information (atime and ctime were 0).

sub read_result
{
  my ($name) = @_;
  my $results = $results{$name};
  my (%pkg,%maintpkg,%sourcepkg,%sourcemax,%arch,$numsub,%release);
  open PKG, "<:utf8","$results" or die "$results not found";
  while(<PKG>)
  {
    my ($type,@values)=split(" ");
    if ($type eq "Package:")
    {
          my @votes = @values;
          my $name = shift @votes;
          unshift @votes,$votes[0]+$votes[1]+$votes[2]+$votes[3];
            $section{$name}='unknown' unless (defined($section{$name}));
            $maint{$name}='Not in sid' unless (defined($maint{$name}));
            $source{$name}='Not in sid' unless (defined($source{$name}));
            for(my $i=0;$i<=$#fields;$i++)
            {
                    my ($f,$v)=($fields[$i],$votes[$i]);
                    $pkg{$name}->{$f}=$v;
                    $maintpkg{$maint{$name}}->{$f}+=$v;
                    $sourcepkg{$source{$name}}->{$f}+=$v;
                    my($sm)=$sourcemax{$source{$name}}->{$f};
                    $sourcemax{$source{$name}}->{$f}=$v 
                      if (!defined($sm) || $sm < $v);
            }
    }
    elsif ($type eq "Architecture:")
    {
      my ($a,$nb)=@values;
      $arch{$a}=$nb;
    }
    elsif ($type eq "Submissions:")
    {
      ($numsub)=@values;
    }
    elsif ($type eq "Release:")
    {
      my ($a,$nb)=@values;
      $release{$a}=$nb;
    }
  }
  close PKG;
  return {'name'      => $name,
          'pkg'       => \%pkg,
          'maintpkg'  => \%maintpkg,
          'sourcepkg' => \%sourcepkg,
          'sourcemax' => \%sourcemax,
          'arch'      => \%arch,
          'release'   => \%release,
          'numsub'    => $numsub};
}

sub gen_sections
{
  my ($stat) = @_;
  my $name = $stat->{'name'};
  my %pkg = %{$stat->{'pkg'}};
  my %maintpkg = %{$stat->{'maintpkg'}};
  my %sourcepkg = %{$stat->{'sourcepkg'}};
  my %sourcemax = %{$stat->{'sourcemax'}};
  my %arch = %{$stat->{'arch'}};
  my %release = %{$stat->{'release'}};
  my $numsub = $stat->{'numsub'};
  my $popcon = "$popbase$popcon{$name}";
  my $popfile = $popfile{$name};
  my @pkgs=sort keys %pkg;
  my %sections = map {$section{$_} => 1} keys %section;
  my @sections = sort keys %sections;
  my @maints= sort keys %maintpkg;
  my @sources= sort keys %sourcepkg;
  my %winner = ();
  my ($sec, $dir, $f);
  for $sec (@sections)
  {
    my @list = grep {$section{$_} eq $sec} @pkgs;
    make ($popcon, $sec, \%pkg, \%winner, \@list);
  }
  #There is a hack: '.' is both the current directory and
  #the catchall regexp.
  for $sec (".",@dists)
  {
    my @list = grep {$section{$_} =~ /^$sec/ } @pkgs;
    make ($popcon, $sec, \%pkg, \%winner, \@list);
  }
  make ($popcon, "maint", \%maintpkg, \%winner, \@maints);
  make ($popcon, "source", \%sourcepkg, \%winner, \@sources);
  make ($popcon, "sourcemax", \%sourcemax, \%winner, \@sources);

  for $sec (@dists)
  {
    open HTML , ">:utf8", "$popcon/$sec/index.html";
    opendir SEC,"$popcon/$sec";
    popconintro($name,"$sec/index.html");
    printf HTML ("<p>Statistics for the section %-16s sorted by fields: ",$sec);
    print_by (".",$_) for (@fields);
    print HTML ("\n </p> \n");
    printf HTML ("<p> <a href=\"first.html\"> First packages in subsections for each fields </a>\n");
    printf HTML ("<p>Statistics for subsections sorted by fields\n <pre>\n");
    for $dir (sort readdir SEC)
    {
      -d "$popcon/$sec/$dir" or next;
      $dir !~ /^\./ or next;
      printf HTML ("%-16s : ",$dir);
      print_by ($dir,$_) for (@fields);
      print HTML ("\n");
    }
    print HTML ("\n </pre>\n");
    htmlfooter $numsub;
    closedir SEC;
    close HTML;
  }
  for $sec (@dists)
  {
    open HTML , ">:utf8", "$popcon/$sec/first.html";
    opendir SEC,"$popcon/$sec";
    popconintro($name,"$sec/first.html");
    printf HTML ("<p>First package in section %-16s for fields: ",$sec);
    for $f (@fields)
    {
            print_pkg $winner{"$sec/$f"};
    }
    print HTML ("\n </p> \n");
    printf HTML ("<p> <a href=\"index.html\"> Statistics by subsections sorted by fields </a>\n");
    printf HTML ("<p>First package in subsections for fields\n <pre>\n");
    printf HTML ("%-16s : ","subsection");
    for $f (@fields)
    {
            printf HTML ("%-20s ",$f);
    }
    print HTML ("\n","_"x120,"\n");
    for $dir (sort readdir SEC)
    {
            -d "$popcon/$sec/$dir" or next;
            $dir !~ /^\./ or next;
            printf HTML ("%-16s : ",$dir);
            for $f (@fields)
            {
                    print_pkg $winner{"$sec/$dir/$f"};
            }
            print HTML ("\n");
    }
    print HTML ("\n </pre>\n");
    htmlfooter $numsub;
    closedir SEC;
    close HTML;
  }
  open HTML , ">:utf8", "$popcon/index.html";
  popconintro($name,"index.html");
  printf HTML ("<p>Statistics for the whole archive sorted by fields: <pre>");
  print_by (".",$_) for (@fields);
  print HTML ("</pre>\n </p> \n");
  printf HTML ("<p>Statistics by maintainers sorted by fields: <pre>");
  print_by ("maint",$_) for (@fields);
  print HTML ("</pre>\n </p> \n");
  printf HTML ("<p>Statistics by source packages (sum) sorted by fields: <pre>");
  print_by ("source",$_) for (@fields);
  print HTML ("</pre>\n </p> \n");
  printf HTML ("<p>Statistics by source packages (max) sorted by fields: <pre>");
  print_by ("sourcemax",$_) for (@fields);
  print HTML ("</pre>\n </p> \n");
  printf HTML ("<p>Statistics for sections sorted by fields\n <pre>\n");
  for $dir ("main","contrib","non-free","non-US","unknown")
  {
    -d "$popcon/$dir" or next;
    $dir !~ /^\./ or next;
    if ($dir eq "unknown")
    {
      printf HTML ("%-16s : ",$dir);
    }
    else
    {
      printf HTML ("<a href=\"$dir/index.html\">%-16s</a> : ",$dir);
    }
    print_by ($dir,$_) for (@fields);
    print HTML ("\n");
  }
  print HTML  <<'EOF';
</pre>
<table border="0" cellpadding="5" cellspacing="0" width="100%">
<tr>
<td>
Statistics per Debian architectures:
<pre>
EOF
    for $f (grep { $_ ne 'unknown' } sort keys %arch)
    {
      my ($port)=split('-',$f);
      $port="$port/";
      $port="kfreebsd-gnu/" if ($port eq "kfreebsd/");
      printf HTML "<a href=\"http://www.debian.org/ports/$port\">%-16s</a> : %-10s <a href=\"stat/sub-$f.png\">graph</a>\n",$f,$arch{$f};
    }
  if (defined $arch{"unknown"}) {
    printf HTML "%-16s : %-10s <a href=\"stat/sub-unknown.png\">graph</a>\n","unknown",$arch{"unknown"}
  }
  print HTML "</pre></td>\n";
  print HTML  <<'EOF';
<td>
<table>
  <tr><td>
    <img alt="Graph of number of submissions per architectures"
    width="600" height="400" src="stat/submission.png">
  </td></tr>
  <tr><td>
    <img alt="Graph of number of submissions per architectures (last 12 months)"
    width="600" height="400" src="stat/submission-1year.png">
  </td></tr>
</table>
</td>
EOF
  print HTML  <<'EOF';
</tr><tr><td>
Statistics per popularity-contest releases:
<pre>
EOF
    for $f (grep { $_ ne 'unknown' } sort keys %release)
    {
      my($name) = $f;
      $name = "$f ($popconver{$f})" if (defined($popconver{$f}));
      printf HTML "%-25s : %-10s \n",$name,$release{$f};
    }
  if (defined $release{"unknown"}) {
    printf HTML "%-25s : %-10s \n","unknown",$release{"unknown"};
  }
  print HTML "</pre></td>\n";
  print HTML  <<'EOF';
<td>
  <table>
    <tr><td>
      <img alt="Graph of popularity-contest versions in use"
       width="600" height="400" src="stat/release.png">
    </td></tr>
    <tr><td>
      <img alt="Graph of popularity-contest versions in use (12 last months)"
       width="600" height="400" src="stat/release-1year.png">
    </td></tr>
  </table>
</td>
EOF
  print HTML "</tr></table><p>\n";
  print HTML "<a href=\"$popfile\">Raw popularity-contest results</a>\n";
  htmlfooter $numsub;
  close HTML;
}

sub read_packages
{
  my ($file,$dist);
  for $file ("slink","slink-nonUS","potato","potato-nonUS",
      "woody","woody-nonUS","sarge","etch")
  {
    open AVAIL, "<:utf8", "$file.sections" or die "Cannot open $file.sections";
    while(<AVAIL>)
    {
      my ($p,$sec)=split(' ');
      defined($sec) or last;
      chomp $sec;
      $sec =~ m{^(non-US|contrib|non-free)/} or $sec="main/$sec";
      $section{$p}=$sec;
      $maint{$p}="Not in sid";
      $source{$p}="Not in sid";
    }
    close AVAIL;
  }
  mark "Reading legacy packages...";
  for $dist ("stable", "testing", "unstable")
  {
    for (glob("$mirrorbase/dists/$dist/*/binary-*/Packages.gz"))
    {
      /([^[:space:]]+)/ or die("incorrect package name");
      my $file = $1;#Untaint
        open AVAIL, "-|:encoding(UTF-8)","zcat $file";
      my $p;
      while(<AVAIL>)
      {
        /^Package: (.+)/  and do {$p=$1;$maint{$p}="bug";$source{$p}=$p;next;};
        /^Version: (.+)/ && $p eq "popularity-contest" 
          and do { $popver{$dist}=$1; next;};
        /^Maintainer: ([^()]+) (\(.+\) )*<.+>/
          and do { $maint{$p}=join(' ',map{ucfirst($_)} split(' ',lc $1));next;};
        /^Source: (\S+)/ and do { $source{$p}=$1;next;};
        /^Section: (.+)/ or next;
        my $sec = $1;
        $sec =~ m{^(non-US|contrib|non-free)/} or $sec="main/$sec";
        $section{$p}=$sec;
      }
      close AVAIL;
    }
  }
  mark "Reading current packages...";
  for $dist ("stable", "testing", "unstable")
  {
    my($v)=$popver{$dist};
    $popconver{$v}=defined($popconver{$v})?"$popconver{$v}/$dist":$dist;
  }
}

# Main code

read_packages();

mark "Reading packages...";

my %stat = ('all' => read_result('all'),
            'stable' => read_result('stable'));

mark "Reading stats...";

for (keys %stat)
{
  gen_sections($stat{$_});
}

mark "Building pages";
